### ISSUE TITLE
Is running thumbor multiprocess any helpful?

### ISSUE THREAD

**Issue by kkopachev** [MinimalCompact/thumbor#14]
Is there any benefits of running circus inside container to be able to run multiple thumbor processes per container? 
It's usually easy enough to run more containers with single thumbor in each.
That seems to add some complexity to the config. 

I suspect it might help in cases where thumbor has blocking operation happening on main tornado thread and not on thread pool, but that should be identified and fixed in thumbor itself.

Any opinions here?

**Comment by gingerlime** [comment]
@kkopachev I did a few simple benchmarks trying to compare performance with/without circus and can't really detect a major shift in both CPU and I/O bound operations. So I think in terms of performance at least, using circus or multiple docker containers is comparable. Of course, in terms of complexity, there is an added complexity having circus. But there's also an element of convenience that I mentioned. So I think it's worth keeping. I don't have very firm opinion about this though, so I'm happy to discuss pro/cons and perhaps different approaches.

**Comment by DonMartin76** [comment]
I have another little bit regarding why it might make sense to keep circus, and that's the caching disk. We use a dynamically provisioned disk (in Kubernetes) for caching, per container, and thus the caching space is shared among the Thumbor instances in the container. We have scaled in both directions, we run 8 thumbors in a single pod, and two pods for a total of 16 thumbor instances.

(Hello and thanks for publishing this Docker image!)

**Comment by kkopachev** [comment]
@DonMartin76 What kind of caching do you use it for? Is it result storage or storage (loader phase)?

**Comment by DonMartin76** [comment]
Result storage in this case.

**Comment by gingerlime** [comment]
@DonMartin76 thanks for the input! I’ve implemented result storage caching via nginx-proxy in MinimalCompact. I’m curious about your setup and using circus for it. Would you mind sharing a bit more info? Also the Kubernetes recipes could definitely help with some more examples if you’re up for it. I’m actually particularly interested in this caching use-case you mentioned so would be great if you could shed more light on it :)

**Comment by DonMartin76** [comment]
I have to look at the nginx-proxy solution you have in place, I have ignored that up to now. 

What we are doing is the following:

* Each Kubernetes Pod has 8 instances of Thumbor
* A pod has a dynamically provisioned persistent volume which it uses as result storage
* The actual storage is another mounted (but not that fast) storage, based on Azure Files
* We don't use the loader phase, just read the files from a mounted storage
* Additionally, we are not using the uploading functionality of Thumbor, we do that via another service, as we have some special needs regarding authentication and authorization
* If a pod is restarted, it obviously loses its result storage, but it's usually not that bad; images are just a little slower for a while
* As we currently run two pods in parallel, we have double load balancing: Kubernetes internal service proxy, and circus internally in the pods

In short, we are only using Thumbor as an image resizing and cropping engine, and hide it behind another service. This will not be the standard use case of Thumbor, but it works fairly well, and we get internal response times of around 2-5 ms for cached images, which is good enough for us. If such a setup works for you is obviously also a matter of your use case. Ours is to serve the same images to many people in different sizes (responsive website, and thumbnails), and it works quite well.

Does this help? If you want, I can provide a set of YAML files which show how this can be set up using `StatefulSet`. But it might still be Azure specific.

**Comment by ryantxu** [comment]
> If you want, I can provide a set of YAML files which show how this can be set up using `StatefulSet`. But it might still be Azure specific.

yes please!  even if it is Asure specific, it is super helpful to have working examples.  We could put it in:  https://github.com/MinimalCompact/thumbor/tree/master/recipes/kubernetes  with 'asure' in the name

**Comment by kkopachev** [comment]
Hm, would you achieve same caching benefit if each container was running just one instance of thumbor with bigger thread pool?

**Comment by gingerlime** [comment]
@kkopachev can you explain how to do this with thumbor? (control the thread pool)

@DonMartin76 I'm also curious about the specifics :) I'm not sure how you configure circus exactly either, so I'd like to learn more if you don't mind sharing.

**Comment by DonMartin76** [comment]
@gingerlime @kkopachev I am also curious about the thread pool thing. This is what I was expecting when I saw Thumbor the first time, and then I was seeing that it works on each request after the other, i.e. serialized, when we tried to display a page with many images.

Regarding the Kubernetes StatefulSet thing, I will try to create a PR against the docs, hope I have time later.

**Comment by kkopachev** [comment]
Thumbor is based on Tornado, event loop framework. For such approach (event loop) to work efficiently, code has to be minimal and all IO operations should be happening outside of the loop. So when thumbor makes request to loader to get the file from somewhere over the network, thumbor can do accept and process another request. All blocking operations - IO and heavy computations - must be happening outside of the event loop, so it can process another requests.

Most of operations are already async in thumbor. That includes loaders, storage, result storage.
What's left is image operations, which are fairly heavy and time consuming. In order to make them async (non blocking), there is a concept of ThreadPool where operation can be offloaded. It's basically a pool of threads (surprise!) which does what you ask it to do and signals back when finished. Number of threads that pool has is controlled by config var ENGINE_THREADPOOL_SIZE. 

So in theory, it all should just work automatically without spawning multiple thumbor processes.
Not sure if threads are spread across CPU cores, though.
ThreadPool still subject to [global interpreter lock](https://docs.python.org/3.6/glossary.html#term-global-interpreter-lock) so that only one piece of python code runs across all threads. However, Pillow engine releases GIL as much as possible when doing actual image operations, so that should be fine.

**Comment by gingerlime** [comment]
Thanks @kkopachev. Really interesting. It looks like it [defaults to 0 though](https://github.com/thumbor/thumbor/blob/b4096a957b8ea2420ea2d91818129dbc482c1ca6/thumbor/thumbor.conf#L129). I guess it makes sense if you don't know the number of available cpus.

From my experience with things like [Puma](http://puma.io/) (ruby), I think typically people use processes *and* threads. So intuitively, I'm not absolutely sure threads alone are good-enough, but it's just speculation. I also think the GIL can become a factor, and processes work around it by launching multiple processes (again, my knowledge is rather limited here). 

I'd be curious to try to benchmark this against processes. I did some quick benchmark between circus and docker processes, and couldn't really spot any meaningful differences, but maybe my benchmark wasn't good enough(?)... I think some real-world benchmarks and optimization techniques can hopefully bring out some good rules of thumb for how to squeeze the best performance and concurrency out of thumbor. Hope you guys are interested collaborating on this?

**Comment by gingerlime** [comment]
Het @kkopachev 

Finally found a bit of time for this :) 

I've created a branch/PR with some benchmarks that I was running (See #31). I was running it on a VPS with 6 virtual cores on Digital Ocean, and therefore kept the maximum number of users to 200 (I think beyond that performance will just degrade... It peaks at around 180 req/s)

The benchmark is very simple with only two thumbor operations, but it can be easily extended to cover more use-cases, loads etc.

For the first benchmark I was trying to compare using `ENGINE_THREADPOOL_SIZE` vs Docker processes vs Thumbor processes inside a single docker container.

I didn't go over the results in too much detail, but at a glance, it seems like multiple processes (either using more docker processes or more thumbor processes inside) reach higher concurrency and perform better overall.

Would be great if you could take a look. I'll try to attach the benchmark log + csv files so you can see the actual data. Would love to continue this conversation, create more benchmarks etc.

**Comment by gingerlime** [comment]
I ran another benchmark, this time with `simd-sse4` and see similar patterns. Latency is lower, but using multiple processes seems to increase throughput more than multiple threads (perhaps I was using too few threads though? not sure what's the optimal number there...)

Here's a [Google sheet with consolidated results](https://docs.google.com/spreadsheets/d/1x0MRSavozv0rsMgcko9tqh2nW8BTQOMHvKM1vojGnW4/edit?usp=sharing)

**Comment by kkopachev** [comment]
Such great results.
Kind of sad that using threads does not help as much as I hoped for.
Wonder what's causing threads to perform that bad. Probably it is still blocking in a lot of places, where it should just call async and handle another request. It supposed to use thread pool only for blocking image operations. All storage and loader operations should be already async. 
Great that you have code to run it, so we can see how newer versions of thumbor perform.

Since you run it on a machine with 6 cores, wonder why it didn't degrade running 9 processes.

**Comment by gingerlime** [comment]
@kkopachev thanks.

> Wonder what's causing threads to perform that bad. Probably it is still blocking in a lot of places, where it should just call async and handle another request. It supposed to use thread pool only for blocking image operations. All storage and loader operations should be already async.

Yes, I was also hoping that it would work "out-of-the-box"... I can't really comment because I'm not that familiar with thumbor's internals unfortunately. The benchmark only measured two thumbor operations on one particular image, so I'm not sure it reflects thumbor as a whole, but it's a starting point.

> Great that you have code to run it, so we can see how newer versions of thumbor perform.

Yes, I hope to use it in future. I tested quickly on an older version of thumbor (6.5.2) and saw similar results. Again, I think if we want to use the benchmark as a more robust tool, we should create more realistic "scenarios" with more operations than just crop and `fit-in` on one image... The good thing is that we now have the foundation to build on top of. cc @heynemann 

> Since you run it on a machine with 6 cores, wonder why it didn't degrade running 9 processes.

I was also wondering, but I think there's still some I/O involved (network, tmp files being saved), so with 6 processes, you still don't saturate all cores completely. I think it's a fairly common rule-of-thumb to use something like 1.5-2x number of cores for processes in a multi-process system (?)... perhaps it's an urban myth than something solid though?
